utils.go

```diff
@@ -82,25 +83,39 @@ func GetRootlessRuntimeDir(rootlessUID int) (string, error) {
 	return path, nil
 }
 
+func isRootlessRuntimeDirOwner(dir string) bool {
+	st, err := system.Lstat(dir)
+	return err == nil && int(st.UID()) == os.Getuid() && st.Mode()&0700 == 0700 && st.Mode()&0066 == 0000
+}
+
 func getRootlessRuntimeDir(rootlessUID int) (string, error) {
 	runtimeDir, err := homedir.GetRuntimeDir()
 	if err == nil {
 		return runtimeDir, nil
 	}
-	tmpDir := fmt.Sprintf("/run/user/%d", rootlessUID)
-	st, err := system.Stat(tmpDir)
-	if err == nil && int(st.UID()) == os.Getuid() && st.Mode()&0700 == 0700 && st.Mode()&0066 == 0000 {
-		return tmpDir, nil
+	initCommand, err := ioutil.ReadFile("/proc/1/comm")
+	if err != nil || string(initCommand) == "systemd" {
+		tmpDir := fmt.Sprintf("/run/user/%d", rootlessUID)
+		if isRootlessRuntimeDirOwner(tmpDir) {
+			return tmpDir, nil
+		}
 	}
-	tmpDir = fmt.Sprintf("%s/%d", os.TempDir(), rootlessUID)
-	if err := os.MkdirAll(tmpDir, 0700); err != nil {
-		logrus.Errorf("failed to create %s: %v", tmpDir, err)
-	} else {
+	tmpDir := fmt.Sprintf("%s/containers-user-%d", os.TempDir(), rootlessUID)
+	if _, err := os.Stat(tmpDir); os.IsNotExist(err) {
```

Comment:
We need to know whether directory exist and using `Stat`. What is the problem with it?

Reply:
Well, that it doesn’t _do so_. This is checking whether the target of a symlink exists.

Sure, the code ends up doing the right thing because the following code on both branches rejects symlinks, but one shouldn’t need to do such analysis to verify security-critical code is correct when there is a readily available alternative — discussed in this PR several times already — that locally does the right thing without extra reasoning.

Reply:
> This is checking whether the target of a symlink exists.

Yes, what is the security problem with symlink?

Reply:
> Sure, the code ends up doing the right thing… but…

Reply:
Please describe steps one-by-one to reproduce attack. I will convert this steps into actual code. Because now I can't understand what you want to say. I see no problem with symlinks.

Reply:
Dan provided good solution to use Mkdir, so resolved.

Reply:
This is unrelated to the `Rename` conversation.